File: examples/ruby/receive_webhook.rb
# frozen_string_literal: true

# Minimal WEBrick server that receives Pushfy webhooks and verifies the
# signature against the RAW request body via Pushfy::Webhooks.
#
#   PUSHFY_WEBHOOK_SECRET=... ruby receive_webhook.rb
#   # then POST to http://localhost:4567/webhooks/messaging
#
# Env vars:
#   PUSHFY_WEBHOOK_SECRET  Shared secret for the messaging webhook (required)
#   PORT                   Listen port (default 4567)
#
# Signature schemes (see Pushfy::Webhooks):
#   - messaging / push -> "sha256=<hex>"  (header X-Pushfy-Signature / X-Push-Signature)
#   - conversations    -> raw "<hex>"     (header X-PA-Signature)

require "pushfy"
require "webrick"
require "json"

def build_server
  secret = ENV.fetch("PUSHFY_WEBHOOK_SECRET")
  port = Integer(ENV.fetch("PORT", "4567"))

  server = WEBrick::HTTPServer.new(Port: port)

  server.mount_proc "/webhooks/messaging" do |req, res|
    # Verify against the exact bytes received — never a re-serialized object.
    raw_body = req.body.to_s
    signature = req["X-Pushfy-Signature"]

    ok = Pushfy::Webhooks.messaging(
      payload: raw_body,
      signature: signature,
      secret: secret
    )

    unless ok
      res.status = 401
      res.body = "invalid signature"
      next
    end

    # Respond fast (2xx), then process asynchronously in real code.
    event = begin
      JSON.parse(raw_body)
    rescue JSON::ParserError
      {}
    end
    puts "Verified webhook: #{event.inspect}"

    res.status = 200
    res.body = "ok"
  end

  trap("INT") { server.shutdown }
  server
end

if $PROGRAM_NAME == __FILE__
  begin
    server = build_server
    puts "Listening on http://localhost:#{server.config[:Port]}/webhooks/messaging"
    server.start
  rescue KeyError => e
    warn "Missing environment variable: #{e.message}"
    exit 1
  end
end

File: examples/node/receive-webhook.js
'use strict';

// Minimal Express server that receives Pushfy delivery-status (DLR) webhooks
// and verifies the HMAC signature before trusting the payload.
//
// Run:  WEBHOOK_SECRET=... node receive-webhook.js
//   then point your Pushfy webhook at  http://<host>:3000/webhooks/pushfy
//
// Requires express:  npm install express

const express = require('express');
const { Pushfy } = require('@pushfy/pushfy');

const app = express();

// IMPORTANT: keep the RAW body. Signature verification hashes the exact bytes
// received — a re-serialized JSON object would produce a different signature.
app.use(express.raw({ type: '*/*' }));

app.post('/webhooks/pushfy', (req, res) => {
  // Messaging status webhooks send:  X-Pushfy-Signature: sha256=<hex>
  const ok = Pushfy.webhooks.messaging({
    payload: req.body,                          // raw Buffer
    signature: req.header('X-Pushfy-Signature'),
    secret: process.env.WEBHOOK_SECRET,
  });

  if (!ok) {
    console.warn('Rejected webhook: bad signature');
    return res.sendStatus(401);
  }

  // Respond fast (200), then process asynchronously so Pushfy doesn't retry.
  res.sendStatus(200);

  const event = JSON.parse(req.body.toString('utf8'));
  setImmediate(() => {
    console.log('DLR received:', event);
    // e.g. update your DB by ext_id / status here
  });
});

const port = process.env.PORT || 3000;
app.listen(port, () => console.log(`Listening for webhooks on :${port}`));
